modules/reference_density.py

```diff
@@ -0,0 +1,73 @@
+# coding=utf-8
+"""
+Created on 3.1.2022
+Updated on 3.1.2022
+
+Potku is a graphical user interface for analyzation and
+visualization of measurement data collected from a ToF-ERD
+telescope. For physics calculations Potku uses external
+analyzation components.
+Copyright (C) 2022 Joonas Kopoonen and Tuomas Pitkänen
+
+This program is free software; you can redistribute it and/or
+modify it under the terms of the GNU General Public License
+as published by the Free Software Foundation; either version 2
+of the License, or (at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program (file named 'LICENCE').
+"""
+__author__ = "Joonas Koponen \n Tuomas Pitkänen"
+__version__ = "2.0"
+
+import warnings
+from typing import List
+
+from modules.general_functions import convert_amu_to_kg
+
+
+class ReferenceDensity:
+    """Recalculates the reference density when the user makes changes to the
+    layer properties on the target composition tab on the simulation side.
+    """
+    def __init__(self, layers: List = None):
+        """Initializes a ReferenceDensity object.
+
+        Args:
+            layers: The reference density will be updated on these element
+            layers (such as Si, Au etc.)
+        """
+
+        self.layers = layers
+        self.reference_density = 0.0
+        self.thickness_limit = 10.0
+        # Total thickness of the layers that are included in the calculation
+        self.total_thickness = 0.0
+
+    def update_reference_density(self):
+        if len(self.layers) == 0:
+            warnings.warn("There are no layers!")

```

Comment:
`warnings.warn()` isn't visible to the end user.

Reply:
OK

Reply:
Is this fine?

```
if not self.layers:
    if 'unittest' in sys.modules:
        warnings.warn("There are no layers!")
        return
    else:
        QtWidgets.QMessageBox.critical(self,
                                       "Warning",
                                       "There are no layers!",
                                       QtWidgets.QMessageBox.Ok,
                                       QtWidgets.QMessageBox.Ok)
        return
```

Reply:
First, you'll have to decide whether this is important enough to display to the user or not. If it is, maybe raise an error, catch it in the UI code and create a dialog there. If it isn't, just remove the `warnings.warn()` and keep the `return` statement.

I don't think the user needs to see this. Otherwise the dialog would show up when a new simulation is created, I think. That would get annoying.